File: Matlab/DeepGLM/nnfun/nnGradLogLikelihood.m
function [gradient,nnOut] = nnGradLogLikelihood(W_seq,beta,X,y,datasize,distr,mean_sigma2_inverse)
%NNGRADIENTLLH Calculate gradient of log likelihood
%   Detailed explanation goes here
%
%   Copyright 2018 Minh-Ngoc Tran (minh-ngoc.tran@sydney.edu.au) and Nghia
%   Nguyen (nghia.nguyen@sydney.edu.au)
%   
%   http://www.xxx.com
%
%   Version: 1.0
%   LAST UPDATE: April, 2018

n = length(y);
[back_prop,nnOut] = nnBackPropagation(X,y,W_seq,beta,distr);
nnOut = nnOut';
switch distr
    case 'normal'
        gradient_theta = mean_sigma2_inverse*back_prop;
        gradient = datasize/n*gradient_theta;   % To compensate the variation
    case 'binomial'
        gradient = datasize/n*back_prop;
    case 'poisson'
        gradient = datasize/n*back_prop;
end
end

File: R/02_libs/nnGradLogLikelihood.R
# Function to calculate gradient of log-likelihood
nnGradLogLikelihood <- function(w_seq,beta,X,y,datasize,distr,mean_sigma2_inverse){
  output <- list()
  n = nrow(X)
  out <- nnBackPropagation(X,y,w_seq,beta,distr)
  back_prop <- out$gradient
  nnOut <- t(out$nnOut)

  switch (distr,
          normal = {gradient_theta <- mean_sigma2_inverse*back_prop
          gradient <- datasize/n*gradient_theta  },        # To compensate the variational lowerbound
          binomial = {gradient <- datasize/n*back_prop},
          poisson = {gradient <- datasize/n*back_prop},
          default)
  output$gradient <- gradient
  output$nnOut <- nnOut
  return(output)
}
